pubsub/gcloud/aio/pubsub/utils.py

```diff
@@ -0,0 +1,46 @@
+import asyncio
+import concurrent
+import threading
+
+import google.api_core.future
+
+def convert_google_future_to_concurrent_future(
+        future: google.api_core.future.Future, *,
+        loop: asyncio.AbstractEventLoop) -> None:
+    """
+    The google-cloud-pubsub subscription library returns a
+    `google.cloud.pubsub_v1.subscriber.futures.StreamingPullFuture`,
+    which is a subclass of `google.api_core.future.Future`, which in turn
+    is NOT a subclass of `concurrent.futures.Future` (even though it is
+    explicitly designed to be interface-identical).
+
+    A `concurrent.futures.Future` can be added to an asyncio task queue
+    with `asyncio.wrap_future`, but that method explicitly calls
+    `isinstance` rather than duck-typing the future.
+
+    This method exists as a hack to make `asyncio.wrap_future` think that
+    a google `Future` is valid.
+
+    Here are the gotchas is uses to do so:
+    - sets `future.__class__` so the `isinstance` check works
+    - sets `future._condition`, `future._state`, and `future._done_callbacks`
+      to their equivalent expected values (these are the attributes which
+      Google decided to avoid mirroring from `concurrent.futures`)
+    - spawns an infinite task which `await`s every second, which prevents the
+      Google future from occasionally getting stuck
+    """
+    # BEWARE: here be dragons
+    async def await_on_interval(interval: int) -> None:
```

Comment:
**COPIED FROM PREVIOUS PR**
**egalpin** 3 days ago
Is this function a good idea? What's the run-time story? The task wakes every 1 second? Can it be preempted if its underlying IO task completes before the next tick? Can a failure mode result in a task that just loops forever?

We've used it before in offline from the sounds of it. I just mean...how much faith do we have in this yielding consistent behavior?

Reply:
@TheKevJames do you have any additional incite here?

Reply:
To be honest, this is mostly magic and dragons to me, as well. Here's what I remember:
- Google's subscriber would never wake up after the first task unless I had _any_ other async task scheduled at all times -- eg. running a task which completes didn't do the trick.
- If the interval was sub-second, profiling suggested a huge amount of time (think >90%) was spent in that function.
- If the interval was large, subscriptions would get batched, ie. we'd grab `flow_control.max_messages` and process them, then do nothing until the next time this method woke up.

So basically my thought is that fixing Google subscriptions requires a noop async looping task which does nothing but kick the scheduler as often as possible but no more often than about a second. That... sounds idiotic to write out, so I suspect we could tune this a bit more.